### ISSUE TITLE
Packing bundled JRE into extecutable

### ISSUE THREAD

**Issue by theexiile1305** [TheBoegl/gradle-launch4j#86]
**I've got the following problem:** The clients of my application don't have the rights to install or simply do not have any JRE installed on their computers. 
A nice feature would be to let the executable have a bundled JRE inside. The distribution of java application is then less complex because the clients don't have to install a JRE.
This issue is related to #64 and launch4j is still able to bundle the JRE inside the executable and only the plugin have to implement this.

I can create a pull request for this improvement. What do you think @TheBoegl?

**Comment by TheBoegl** [comment]
I'm not aware, that such a feature exists in launch4j.
Please provide the link to it's documentation.

If your users aren't allowed to install in user-only mode, you could provide a archive containing your created executable and jre.

**Comment by theexiile1305** [comment]
Thank you for your fast response 😄 
You can include any JRE inside the executable. See  http://launch4j.sourceforge.net/docs.html and in the example configuration file there are the following lines:

```
<jre>
    ...
    <path>bundled JRE path (%VAR%)</path>
    ...
</jre>
```

Right, my users don't have admin rights and they want one executable without any dependencies on the disk, like a single executable on the desktop.

**Comment by TheBoegl** [comment]
This already works as described. Just use the property `bundleJrePath` though you will still need to provide your users with a non administrative installation (e.g in [innosetup](http://www.jrsoftware.org/ishelp/index.php?topic=admininstallmode)).
With that you could create a desktop or start menu entry pointing to your install location.

**Comment by theexiile1305** [comment]
I have only the following configuration on a mac and it doesn't work:
```
<jre>
    ...
    <path>/Users/user/Downloads/jre</path>
    ...
</jre>
```
What's the failure?

**Comment by TheBoegl** [comment]
Please provide an example of your gradle configuration and state what you think should happen compared to what happens.

**Comment by theexiile1305** [comment]
The following example as the configuration is given and I execute the Gradle task `createExe` on my mac: 
```
launch4j {
    mainClassName = 'org.springframework.boot.loader.JarLauncher'
    headerType = 'console'
    icon = "${projectDir}/launch4j/logo.ico"
    outfile = 'MyFancyProgram.exe'
    windowTitle = 'MyFancyProgram'
    productName = 	'MyFancyProgram'
    fileDescription = 'MyFancyProgram'
    language = 'ENGLISH_US'
    bundledJrePath = "/Users/user/Downloads/jdk"
}
```

**What should happen:**
I want to create a standalone executable with a JRE inside the executable. Therefore the user doesn't have to install the JRE by himself. 
**What happens:**
The executable was successfully created but on startup, the following failure message appears:

**Comment by TheBoegl** [comment]
You expect the executable to work in another way as stated in the section jre of [launch4j's documentiation](http://launch4j.sourceforge.net/docs.html)
```
<path>, <minVersion>, <maxVersion>
The <path> property is used to specify the absolute or relative path (to the executable) of a bundled JRE, it does not rely on the current directory or <chdir>. Note that this path is not checked until the actual application execution. If you'd like the wrapper to search for a JRE (public or SDK private) use the <minVersion> property, you may also specify the <maxVersion> to prevent it from using higher Java versions. Launch4j will always use the highest version available (in the min/max range of course). If an Oracle's JRE is not available or does not satisfy the search criteria, the search will be repeated on IBM runtimes. You can also combine these properties to change the startup process...
```
As stated [above](/TheBoegl/gradle-launch4j/issues/86#issuecomment-514158551): You have to provide the jre to your users as there is no other way. (either you use the suggested innosetup settings or install4j or a simple zip file as suggested by Francisco Puga in this [so answer](https://stackoverflow.com/a/7582064))
In other words: the jre is not and will never be packed into the executable.

**Comment by theexiile1305** [comment]
That's not correct.
If I use the following launch4j configuration the JRE is bundled into the executable and on a blank new machine, nobody has to install any JRE or JDK by themselves. 
```
<?xml version="1.0" encoding="UTF-8"?>
<launch4jConfig>
    <dontWrapJar>false</dontWrapJar>
    <headerType>console</headerType>
    <jar>... JAR PATH ...</jar>
    <outfile>... OUTPUT PATH ...</outfile>
    <errTitle></errTitle>
    <cmdLine></cmdLine>
    <chdir>.</chdir>
    <priority>normal</priority>
    <downloadUrl></downloadUrl>
    <supportUrl></supportUrl>
    <stayAlive>false</stayAlive>
    <restartOnCrash>false</restartOnCrash>
    <manifest></manifest>
    <icon>... ICON PATH ...</icon>
    <classPath>
        <mainClass>org.springframework.boot.loader.JarLauncher</mainClass>
    </classPath>
    <jre>
        <path>/Users/user/Downloads/jre</path>
        <bundledJre64Bit>true</bundledJre64Bit>
        <bundledJreAsFallback>false</bundledJreAsFallback>
        <minVersion>1.11</minVersion>
        <maxVersion>1.11</maxVersion>
        <runtimeBits>64</runtimeBits>
    </jre>
    <versionInfo>
        <fileVersion>0.0.0.1</fileVersion>
        <txtFileVersion>free from</txtFileVersion>
        <fileDescription>Fancy Application</fileDescription>
        <copyright>copyright</copyright>
        <productVersion>0.0.0.1</productVersion>
        <txtProductVersion>free from</txtProductVersion>
        <productName>Fancy Application</productName>
        <internalName>Fancy Application</internalName>
        <originalFilename>Fancy Application.exe</originalFilename>
        <trademarks></trademarks>
        <language>ENGLISH_US</language>
    </versionInfo>
</launch4jConfig>
```

**Comment by david-hall** [comment]
A working example of using Launch4J to bundle a Java application with a
specific JRE can be found at
https://github.com/vZome/vzome/blob/master/desktop/platform/windows/README.md.
This particular distribution is configured to require vZome
<http://vzome.com/> to use the bundled JRE rather than any JRE found on the
target platform.

I have also posted this answer on the previously referenced stackoverflow
question
<https://stackoverflow.com/questions/7071133/how-to-bundle-a-jre-with-launch4j/57279837#57279837>
.

Hope this helps.

**Comment by TheBoegl** [comment]
@theexiile1305 please run both build of your app with --l4j-debug option which creates a launch4j log and attach it here.
